core/math/math_funcs.h

```diff
@@ -74,11 +74,23 @@ class Math {
 	static _ALWAYS_INLINE_ double tanh(double p_x) { return ::tanh(p_x); }
 	static _ALWAYS_INLINE_ float tanh(float p_x) { return ::tanhf(p_x); }
 
-	static _ALWAYS_INLINE_ double asin(double p_x) { return ::asin(p_x); }
-	static _ALWAYS_INLINE_ float asin(float p_x) { return ::asinf(p_x); }
+	static _ALWAYS_INLINE_ double safe_asin(double p_x) { return ::asin(CLAMP(p_x, -1, 1)); }
+	static _ALWAYS_INLINE_ float safe_asin(float p_x) { return ::asinf(CLAMP(p_x, -1, 1)); }
```

Comment:
Instead of clamping, it might be more performant to just return a fixed value if above 1.0 and a fixed value if below -1.0. This way those cases can just skip `asin` entirely, and it should have the same amount of branching (CLAMP has up to 2 checks already).

Reply:
Except that CLAMP can be implemented with conditional moves, so these branches may not actually exist. This really depends on the complexity of asin and which cases to optimize for. https://godbolt.org/z/Kd4511oxh

Reply:
Thanks for the Godbolt link. It would be great if someone could benchmark this. Conditional moves are still conditional, so they might still have a performance cost by interrupting the CPU's pipeline (since it can't predict which value will be passed into `asin`).

Reply:
Agree these things are kind of hard to predict depending on CPU / SIMD etc. Benchmarking on linux with `-O3` suggested @aaronfranke 's suggestion was marginally faster, so I've changed for now, but am happy to switch back if evidence goes the other way.

Reply:
For me the branched version is about 15-20% faster, depending on the distribution of numbers used as inputs:
```cpp
// ...
static std::vector<double> ManyValues() {
    std::random_device rd {};
    std::mt19937 gen {rd()};
    std::normal_distribution<> d {0, 1};
    // std::uniform_real_distribution<> d {-1, 1};
    std::vector<double> result;
    for (int i = 0; i < 1000; i++) {
        result.push_back(d(gen));
    }
    return result;
}
static void BenchAsinClamp(benchmark::State& state) {
  std::vector<double> values = ManyValues();
  for (auto _ : state) {
    for (double x : values) {
      double v = asin_clamp(x);
      benchmark::DoNotOptimize(v);
    }
  }
}
// ...
```

Reply:
15-20% faster is definitely a strong argument in favor of the branching over clamping the inputs.